src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java

```diff
@@ -905,14 +960,111 @@ protected void createFlattenedDependencies( Model effectiveModel, List<Dependenc
+        final Set<String> processedDependencies = new HashSet<>();
+
+        final Artifact projectArtifact = this.project.getArtifact();
+
+        final DependencyNode dependencyNode = this.dependencyTreeBuilder.buildDependencyTree(this.project,
+                this.localRepository, null);
+
+        dependencyNode.accept(new DependencyNodeVisitor() {
+            @Override public boolean visit(DependencyNode node) {
+                if (node.getArtifact().getGroupId().equals(projectArtifact.getGroupId()) && node.getArtifact().getArtifactId().equals(projectArtifact.getArtifactId())) {
+                    return true;
+                }
+                if (node.getState() != DependencyNode.INCLUDED) return true;
+                dependencyNodeStack.push(node);
+                return true;
+            }
+
+            @Override public boolean endVisit(DependencyNode node) {
+                return true;
+            }
+        });
+
+        while (!dependencyNodeStack.isEmpty()) {
+            DependencyNode node = dependencyNodeStack.pop();
+
+            Artifact artifact = node.getArtifact();
+
+            Dependency dependency = new Dependency();
+            dependency.setGroupId(artifact.getGroupId());
+            dependency.setArtifactId(artifact.getArtifactId());
+            dependency.setVersion(artifact.getVersion());
+            dependency.setClassifier(artifact.getClassifier());
+            dependency.setOptional(artifact.isOptional());
+            dependency.setScope(artifact.getScope());
+            dependency.setType(artifact.getType());
+
+            List<Exclusion> exclusions = new LinkedList<>();
+
+            org.eclipse.aether.artifact.Artifact aetherArtifact = new DefaultArtifact(artifact.getGroupId(), artifact.getArtifactId(), null, artifact.getVersion());
+            ArtifactDescriptorRequest request = new ArtifactDescriptorRequest(aetherArtifact, null, null);
+            ArtifactDescriptorResult artifactDescriptorResult = this.artifactDescriptorReader
+                    .readArtifactDescriptor(this.session.getRepositorySession(), request);
+
+            for (org.eclipse.aether.graph.Dependency artifactDependency: artifactDescriptorResult.getDependencies()) {
+                if ("test".equals(artifactDependency.getScope())) {
+                    continue;
+                }
+                Exclusion exclusion = new Exclusion();
+                exclusion.setGroupId(artifactDependency.getArtifact().getGroupId());
+                exclusion.setArtifactId(artifactDependency.getArtifact().getArtifactId());
+                exclusions.add(exclusion);
+            }
+
+            dependency.setExclusions(exclusions);
+
+            // convert dependency to string for the set, since Dependency doesn't implement equals, etc.
+            String dependencyString = dependency.toString();
+
+            if (processedDependencies.contains(dependencyString)) {
```

Comment:
Java's Set returns true if this set did not already contain the specified element. (so that we can shorten such code)

```
if (!processedDependencies.add(dependencyString)) {
  // continue if dependencyString is already checked
  continue;
}
```

Reply:
nice touch! updated.